---
title: "DESeq"
output: html_document
---

```{r}
library("DESeq2")
library("Biostrings")
library("Biobase")
library("dplyr")
library("plyr")
library("stringr")
library("coRdon")
library("data.table")
library("ggplot2")
library("dplyr")
library("pheatmap")
library("RColorBrewer")
library("ggbiplot")
library("pathview")
library("gage")
```

```{r}
my_fasta_file_folder <- "C:\\Users\\Maja\\Desktop\\Eva - diplomski\\samples\\60uzoraka"
#my_fasta_file_folder <- "C:\\Users\\pavli\\Desktop\\FAKS\\DS\\4. semestar\\diplomski rad\\samples\\10uzoraka"
codonsInSamples <- readSet(my_fasta_file_folder, prepend.filenames = T)
codonsInSamples2 <- codonTable(codonsInSamples)
#head(codonsInSamples2)
```

```{r}
#samples <- str_extract(getID(codonsInSamples3), "[HL]D")
samples <- str_extract(getID(codonsInSamples2), "[HL]D\\d+")
#s2 <- substr(samples,1,1)
```

```{r}
melp <- MELP(codonsInSamples2, ribosomal = TRUE)
```


Density plot za 10 ^ meadian(MELP):
```{}
dt <- data.table(sample = samples, genes = getKO(codonsInSamples2), MELP = as.numeric(melp))
dt[, MELP := median(MELP), by = c("sample", "genes")]
ggplot(dt, aes(10^MELP)) + geom_density() + coord_cartesian(xlim = c(0,20))
```


Priprema podataka:
```{r}
dt <- data.table(sample = samples, genes = getKO(codonsInSamples2), MELP = as.numeric(melp))

dtMelp <- copy(dt)

dt[, MELP := round(10^median(MELP)), by = c("sample", "genes")]
dt <- unique(dt)
dt <- dcast(dt, genes ~ sample, value.var = "MELP")
dt[is.na(dt)] <- 0
dt <- as.data.frame(dt)
rownames(dt) <- dt$genes
dt <- dt[, 2:ncol(dt)]
```


DESeq:
```{r}
coldata <- data.frame(sample = unique(samples))
coldata$condition <- "healthy"
coldata$condition[which(substr(coldata$sample, 1, 1) == "L")] <- "cirrhotic"
coldata$condition <- factor(coldata$condition, levels = c("healthy", "cirrhotic"))
dataset <- DESeqDataSetFromMatrix(countData = dt, colData = coldata, design = ~ condition)
#head(assay(dataset))
```

```{r}
#pre-fitlering ne treba raditi jer su prethodno vec uklonjeni redovi koji imaju 0
#dataset <- dataset[ rowSums(counts(dataset)) > 1, ]
#nrow(dataset)
```

Transformation
```{r}
rlmean <- rlog(dataset, fitType = "mean")
#vsd <- vst(dataset)
#rld <- rlog(dataset)
```

```{}
dds <- estimateSizeFactors(dataset)
df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"),
  as_data_frame(assay(rlmean)[, 1:2]) %>% mutate(transformation = "rlog_mean"))
  
colnames(df)[1:2] <- c("x", "y")  
ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) + coord_fixed() + facet_grid( . ~ transformation)  
```


```{r}
distances <- dist(t(assay(rlmean)))
```

```{r}
sampleDistMatrix <- as.matrix(distances)
rownames(sampleDistMatrix) <- unique(samples)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = distances,
         clustering_distance_cols = distances,
         col = colors, 
         cellheight = 6,
         fontsize = 5)
```

PCA plot:
```{}
pcaData <- plotPCA(rlmean, intgroup = "sample", returnData = TRUE)
plotPCA(rlmean, intgroup = "sample")
```

```{r}
rlog_values <- t(rlog(counts(dataset)))
colnames(rlog_values) <- dt$genes
PCA <- prcomp(rlog_values, scale. = TRUE)
ggbiplot(PCA, ellipse = TRUE, var.axes = FALSE, groups = factor(substr(rownames(rlog_values), 1, 1))) + theme_classic()
```

DESeq:
```{r}
analysis <- DESeq(dataset)
res <- results(analysis)
summary(res)
```



Metabolic Module Identification

```{r}
path.set <- kegg.gsets("ko")
ko.gs <- path.set$kg.sets
```

Extraction of fold-change (M) values for the samples.
```{r}
sampleMVals <- as.data.frame(res$log2FoldChange)
```

GAGE run
```{r}
pathwayEnrich <- gage(data.frame(row.names=rownames(res)[res$pvalue < 0.1],res$log2FoldChange[res$pvalue < 0.1]), gsets = ko.gs, samp = NULL, compare = "unpaired", FDR.adj = F)
kegg.sig <- sigGeneSet(pathwayEnrich)
kegg.sig
```









### beskorisni dio


Analiza top gena:
padj < 0,05
LFC > 0, 1, 2
pathview
```{r}
#padj < 0,05
length(which(res$padj < 0.05, arr.ind = T))
res[which(res$padj < 0.05, arr.ind = T), ]
#pvalue < 0,05
length(which(res$pvalue < 0.05, arr.ind = T))
res[which(res$pvalue < 0.05, arr.ind = T), ]
#LFC > 2
length(which(res$lfcSE > 2, arr.ind = T))
res[which(res$lfcSE > 2, arr.ind = T), ]
#LFC > 2 i pvalue < 0,05 
length(which(res$lfcSE > 1 & res$pvalue < 0.05, arr.ind = T))  #41 gen
#topGenes <- res[which(res$lfcSE > 1 & res$pvalue < 0.05, arr.ind = T), ]
#topGenes <- res[which(res$padj < 0.05, arr.ind = T), ]
topGenes4 <- which(res$padj < 0.05, arr.ind = T)
topGenes41 <- res[which(res$lfcSE > 1 & res$pvalue < 0.05, arr.ind = T), ]
top41 <- which(res$lfcSE > 1 & res$pvalue < 0.05, arr.ind = T)
```



Pathview:
```{r}
require(KEGGREST)
paths <- names(keggList("pathway"))
paths <- regmatches(paths, regexpr("{2,4}\\d{5}", paths))
#pnames <- unname(keggList("pathway"))

data("demo.paths")
```

```{r}
setwd("C:\\Users\\Maja\\Desktop\\Eva - diplomski\\pathview")
#setwd("C:\\Users\\pavli\\Desktop\\FAKS\\DS\\4. semestar\\diplomski rad\\R\\pathview")
#data(demo.paths)
pv <- pathview(gene.data = rownames(dt[topGenes41, ]), pathway.id = paths, species = "ko")
pv <- pathview(gene.data = dt[topGenes4, ], pathway.id = paths, species = "ko")
```



```{r}
library(readxl)
table1 <- read_excel("C:\\Users\\Maja\\Desktop\\Eva - diplomski\\60uzoraka.xlsx", col_types = c("text", "text", "text", "numeric", "numeric", "text", "text", "text", "text", "numeric", "numeric", "numeric", "numeric", "numeric", "text", "text", "numeric", "numeric", "text", "text", "guess", "guess"))
table1$`Sample ID` <- paste(str_extract(table1$`Sample ID`, "[HL]D"), str_extract(table1$`Sample ID`, "\\d+"), sep = "")
table1 <- table1[, c(1, 3:9, 11:13)]
#seq_names <- readSet("C:\\Users\\Maja\\Desktop\\Eva - diplomski\\samples\\svi uzorci", prepend.filenames = T)
#imena <- unique(str_extract(names(seq_names), "[HL]D\\d+"))
#selected <- paste(str_extract(imena, "[HL]D"), str_extract(imena, "\\d+"), sep = "-")
#table1 <- table1[table1$`Sample ID` %in% selected, ]
#table1$`Sample ID` <- imena   

```


#######





Korelacije:
```{r}
library(readxl)
table1 <- read_excel("C:\\Users\\Maja\\Desktop\\Eva - diplomski\\rad\\Supplementary information\\nature13568-s1\\Table 1.xlsx", col_types = c("text", "text", "text", "text", "numeric", "numeric", "text", "text", "text", "text", "numeric", "numeric", "numeric", "numeric", "numeric", "text", "text", "numeric", "numeric", "text", "text"))
table1$`Sample ID` <- paste(str_extract(table1$`Sample ID`, "[HL]D"), str_extract(table1$`Sample ID`, "\\d+"), sep = "")
table1 <- table1[1:181, c(2, 4:10, 12:14)]
table1$condition <- "healthy"
table1$condition[which(substr(table1$`Sample ID`, 1, 1) == "L")] <- "cirrhotic"
```

```{r}
proba <- table1[, c(2:4, 12)]
#proba <- table1[, c(2:4, 6:7, 12)]
proba$Gender[proba$Gender == "female"] <- 1
proba$Gender[proba$Gender == "male"] <- 0
proba$condition[proba$condition == "healthy"] <- 1
proba$condition[proba$condition == "cirrhotic"] <- 0
#proba$`HBV related  (Y or N)`[proba$`HBV related  (Y or N)` == "Y"] <- 1
#proba$`HBV related  (Y or N)`[proba$`HBV related  (Y or N)` == "N"] <- 0
#proba$`Alcohol related (Y or N)`[proba$`Alcohol related (Y or N)` == "Y"] <- 1
#proba$`Alcohol related (Y or N)`[proba$`Alcohol related (Y or N)` == "N"] <- 0

#proba <- data.frame(condition = as.numeric(proba$condition), gender = as.numeric(proba$Gender), age = as.numeric(proba$Age), BMI = as.numeric(proba$`BMI (kg/m2)`), HBV_related = as.numeric(proba$`HBV related  (Y or N)`), alcohol_related = as.numeric(proba$`Alcohol related (Y or N)`))

proba <- data.frame(condition = as.numeric(proba$condition), gender = as.numeric(proba$Gender), age = as.numeric(proba$Age), BMI = as.numeric(proba$`BMI (kg/m2)`))


cor(proba)
```





PCA plots:
```{r}
library(readxl)
table1 <- read_excel("C:\\Users\\Maja\\Desktop\\Eva - diplomski\\60uzoraka.xlsx", col_types = c("text", "text", "text", "numeric", "numeric", "text", "text", "text", "text", "numeric", "numeric", "numeric", "numeric", "numeric", "text", "text", "numeric", "numeric", "text", "text"))
table1$`Sample ID` <- paste(str_extract(table1$`Sample ID`, "[HL]D"), str_extract(table1$`Sample ID`, "\\d+"), sep = "")
table1 <- table1[, c(1, 3:8, 11:13)]
```

```{r}
rlog_values <- t(rlog(counts(dataset)))
colnames(rlog_values) <- dt$genes
PCA <- prcomp(rlog_values, scale. = TRUE)
```

```{r}
ggbiplot(PCA, ellipse = F, var.axes = FALSE, groups = factor(table1$Gender)) + theme_classic() + stat_ellipse(aes(group = coldata$condition))
ggbiplot(PCA, ellipse = F, var.axes = FALSE, groups = factor(table1$`HBV related  (Y or N)`)) + theme_classic() + stat_ellipse(aes(group = coldata$condition))
ggbiplot(PCA, ellipse = F, var.axes = FALSE, groups = factor(table1$`Alcohol related (Y or N)`)) + theme_classic() + stat_ellipse(aes(group = coldata$condition))
```

```{r}
BMIs <- cut(table1$`BMI (kg/m2)`, quantile(table1$`BMI (kg/m2)`))
ggbiplot(PCA, ellipse = T, var.axes = FALSE, groups = BMIs) + theme_classic() + stat_ellipse(aes(group = coldata$condition))
```


```{r}
TBs <- factor(round(log(table1$`TB (umol/L)`)*2))
ggbiplot(PCA, choices = c(2,3), ellipse = T, var.axes = FALSE, groups = TBs) + theme_classic() + stat_ellipse(aes(group = coldata$condition))
ggbiplot(PCA, ellipse = T, var.axes = FALSE, groups = TBs) + theme_classic() 
```







Klasteriranje podataka:
```{r}
tablica <- copy(table1)
tablica$Gender[tablica$Gender == "female"] <- 1
tablica$Gender[tablica$Gender == "male"] <- 0
tablica$`HBV related  (Y or N)`[tablica$`HBV related  (Y or N)` == "Y"] <- 1
tablica$`HBV related  (Y or N)`[tablica$`HBV related  (Y or N)` == "N"] <- 0
tablica$`Alcohol related (Y or N)`[tablica$`Alcohol related (Y or N)` == "Y"] <- 1
tablica$`Alcohol related (Y or N)`[tablica$`Alcohol related (Y or N)` == "N"] <- 0
tablica$`Cirrhotic(Y or N)`[tablica$`Cirrhotic(Y or N)` == "Y"] <- 1
tablica$`Cirrhotic(Y or N)`[tablica$`Cirrhotic(Y or N)` == "N"] <- 0
```

```{r}
pairs(apply(as.matrix(tablica[, 2:10]), 2, as.numeric), col = as.numeric(coldata$condition) + 1)
```





S albuminom i bilirubinom:
```{r}
PCA1 <- prcomp(apply(as.matrix(tablica[, c(2:4, 6:10)]), 2, as.numeric), scale. = TRUE)
ggbiplot(PCA1, ellipse = TRUE, var.axes = FALSE, groups = factor(substr(rownames(rlog_values), 1, 1))) + theme_classic()
PCA1$rotation
```


Bez njih:
```{r}
PCA2 <- prcomp(apply(as.matrix(tablica[, c(2:4, 6:8)]), 2, as.numeric), scale. = TRUE)
ggbiplot(PCA2, ellipse = TRUE, var.axes = FALSE, groups = factor(substr(rownames(rlog_values), 1, 1))) + theme_classic()
```

Bez bilirubina:
```{r}
PCA3 <- prcomp(apply(as.matrix(tablica[, c(2:4, 6:9)]), 2, as.numeric), scale. = TRUE)
ggbiplot(PCA3, ellipse = TRUE, var.axes = FALSE, groups = factor(substr(rownames(rlog_values), 1, 1))) + theme_classic()
```

Bez albumina:
```{r}
PCA4 <- prcomp(apply(as.matrix(tablica[, c(2:4, 6:8, 10)]), 2, as.numeric), scale. = TRUE)
ggbiplot(PCA4, ellipse = TRUE, var.axes = FALSE, groups = factor(substr(rownames(rlog_values), 1, 1))) + theme_classic()
```








Funkcije koje klasteriraju i slazu tablice s missclassification rates vrijednostima po klasteru:

Ovo slozi s data table umjesto for petljom:
```{r}
clustering <- function(N){
  kclustering <- kmeans(distances, N)
  classy <- data.table(condition = substr(names(kclustering$cluster), 1, 1), cluster = kclustering$cluster)
  missclass <- data.table()
  for (i in 1:ncol(table(classy))){
    mc_rate <- min(table(classy)[, i]) / sum(table(classy)[, i])
    missclass <- rbind(missclass, data.table(cluster = i, mc_rate))
  }
  missclass$mc_rate
}


mc_table <- function(N){
  missclass <- data.table(cluster = 1:N, 
                          mc_rate = apply(replicate(1000, clustering(N)), 1, sum) / 1000,
                          st_dev = apply(replicate(1000, clustering(N)), 1, sd))
  return(missclass)
}

#for (i in 2:10){
#  print(mc_table(i))
#}
```




#### pretvori u dt

```{r}
#daj prokuzi ovo
kclustering <- kmeans(distances, 7)
classy <- data.table(condition = substr(names(kclustering$cluster), 1, 1), cluster = kclustering$cluster)
missclass <- data.table(cluster = 1:7)
missclass[ , mc_rate := min(.SD) / sum(.SD), by = .I, .SD = table(classy)[, .I]]
missclass[ , mc_rate := min(.SD) / sum(.SD), by = cluster, .SD = table(classy)[, cluster]]
```

####




Funkcije koje odreðuju konsenzus klastere za 1000 klasteriranja:
-ovo bi trebalo biti jedno s onih gornjih 1000 replikacija
```{r}
klasteri <- function(N){
  kclustering <- kmeans(distances, N)
  sample_clusters <- kclustering$cluster
}

consensus_clst <- function(N){
  clusters <- matrix(replicate(1000, klasteri(N)), ncol = 1000)
  clusters <- data.table(cbind(sample = tablica$`Sample ID`, clusters))
  clusters[, final_cluster := which.max(table(as.matrix(.SD))), .SDcols = V2:V1001, by = sample]
  return(clusters$final_cluster)
}
```


Promijeniti broj u mc_table i consensus_clst u zeljeni broj klastera:
```{r}
set.seed(753)
missclass <- mc_table(3)[, 1:2]

#mc_rate ovisnosti nece imati ucinka kada su sve gotovo iste
regresije <- copy(tablica[, c(1, 5, 6, 7, 9, 10)])
colnames(regresije) <- c("sample", "condition", "HBV", "alcohol", "albumin", "bilirubin")
regresije$cluster <- consensus_clst(3)

regresije <- merge(regresije, missclass)
regresije$cluster
```

```{r}
summary(lm(cluster ~ albumin + bilirubin, regresije))
summary(lm(cluster ~ albumin, regresije))
summary(lm(cluster ~ bilirubin, regresije))

summary(lm(cluster ~ condition, regresije))
summary(lm(condition ~ cluster, regresije))

#summary(lm(mc_rate ~ albumin + bilirubin, regresije))
#summary(lm(mc_rate ~ albumin, regresije))

#summary(lm(mc_rate ~ bilirubin, regresije))

#ovisnost klastera o tome je li uzrokovano HBV-om ili ne:
summary(lm(cluster ~ alcohol + HBV, regresije))
summary(lm(cluster ~ HBV, regresije))

#summary(lm(mc_rate ~ alcohol + HBV, regresije))
#summary(lm(mc_rate ~ HBV, regresije))
```



Poveznica direktno izmedu codon usage-a i metadata
```{r}
head(codonsInSamples2)
codon_usage <- data.table(condition = substr(getID(codonsInSamples2), 1, 1),
                          codonCounts(codonsInSamples2))
codon_usage$condition[codon_usage$condition == "H"] <- 1
codon_usage$condition[codon_usage$condition == "L"] <- 0

summary(lm(condition ~ ., codon_usage))

PCA <- prcomp(apply(as.matrix(tablica[, c(2:4, 6:8)]), 2, as.numeric), scale. = TRUE)
ggbiplot(PCA2, ellipse = TRUE, var.axes = FALSE, groups = factor(substr(rownames(rlog_values), 1, 1))) + theme_classic()

PCA <- prcomp(apply(as.matrix(codon_usage), 2, as.numeric), scale. = TRUE)
ggbiplot(PCA, ellipse = TRUE, var.axes = FALSE, groups = factor(substr(codon_usage$condition, 1, 1))) + theme_classic()
```









